File: C#/Container.Tests/IoCContainerTest.cs
using System;
using IoC.Source;
using IoC.Source.Exceptions;
using IoC.Tests.ContainerTestClasses;
using IoC.Tests.ContainerTestClasses.Mock;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace IoC.Tests
{
    [TestClass]
    public class IoCContainerTest
    {
        [TestMethod]
        public void Resolve_NoDependencies_NotRegistered()
        {
            var container = new IoCContainer();
            var res = container.Resolve<NoDependencies>();

            Assert.IsNotNull(res);
        }

        [TestMethod]
        public void Resolve_NoDependencies_Registered()
        {
            var container = new IoCContainer();
            container.RegisterType<NoDependencies, NoDependencies>();
            var res = container.Resolve<NoDependencies>();

            Assert.IsNotNull(res);
        }

        [TestMethod]
        public void Resolve_Interface_Registered()
        {
            var container = new IoCContainer();
            container.RegisterType<IMockService, MockService1>();
            var res = container.Resolve<InterfaceDependency>();

            Assert.IsNotNull(res.Service as MockService1);
        }

        [TestMethod]
        public void Resolve_MultipleDepencencies_NotRegistered()
        {
            var container = new IoCContainer();
            var res = container.Resolve<MultipleDependencies>();

            Assert.IsNotNull(res.Service2);
            Assert.IsNotNull(res.Service1);
        }

        [TestMethod]
        public void Resolve_MultipleDepencencies_AndInterface()
        {
            var container = new IoCContainer();
            container.RegisterType<IMockService, MockService1>();
            var res = container.Resolve<MultipleDependenciesAndInterface>();

            Assert.IsNotNull(res.InterfaceService);
            Assert.IsNotNull(res.InterfaceService is MockService1);
            Assert.IsNotNull(res.Service2);
            Assert.IsNotNull(res.Service1);
        }

        [TestMethod]
        public void Resolve_InjectionAttributeConstructor()
        {
            var container = new IoCContainer();
            var res = container.Resolve<InjectConstructor>();

            Assert.IsNull(res.InterfaceService);
            Assert.IsNotNull(res.Service1);
            Assert.IsNotNull(res.Service2);
        }

        [TestMethod]
        public void Resolve_InjectionAttributeProperty()
        {
            var container = new IoCContainer();
            var res = container.Resolve<InjectProperty>();

            Assert.IsNull(res.InterfaceService);
            Assert.IsNotNull(res.Service1);
            Assert.IsNotNull(res.Service2);
        }

        [TestMethod]
        public void Resolve_Instance()
        {
            var container = new IoCContainer();
            container.RegisterInstance(container.Resolve<NonStaticSingleton>());
            container.Resolve<NonStaticSingleton>();
            Assert.AreEqual(NonStaticSingleton.NConstructed, 1);
            container.Resolve<NonStaticSingleton>();
            Assert.AreEqual(NonStaticSingleton.NConstructed, 1);
            container.Resolve<NonStaticSingleton>();
            Assert.AreEqual(NonStaticSingleton.NConstructed, 1);
        }

        [TestMethod]
        public void Resolve_Func()
        {
            var container = new IoCContainer();
            int n = 0;
            container.RegisterFunc<NoDependencies>(() =>
            {
                n++;
                return new NoDependencies();
            });

            var res1 = container.Resolve<NoDependencies>();
            Assert.IsNotNull(res1);
            Assert.AreEqual(n, 1);
            var res2 = container.Resolve<NoDependencies>();
            Assert.IsNotNull(res2);
            Assert.AreEqual(n, 2);
            var res3 = container.Resolve<NoDependencies>();
            Assert.IsNotNull(res3);
            Assert.AreEqual(n, 3);
        }

        [TestMethod]
        public void Resolve_DoubleSameDependency()
        {
            var container = new IoCContainer();
            container.RegisterType<IMockService, MockService1>();

            var res1 = container.Resolve<DoubleSameDependency>();
            Assert.IsNotNull(res1);
        }


        #region Exceptions
        [TestMethod]
        [ExpectedException(typeof(ArgumentNullException), "Instance may not be null when registering instance.")]
        public void Resolve_NullInstance()
        {
            var container = new IoCContainer();
            container.RegisterInstance<object>(null);
        }

        [TestMethod]
        [ExpectedException(typeof(ArgumentNullException), "Func may not be null when registering instance.")]
        public void Resolve_NullFunc()
        {
            var container = new IoCContainer();
            container.RegisterFunc<object>(null);
        }

        [TestMethod]
        [ExpectedException(typeof(CircularDependencyException), "A class with a circular dependency was trying to be resolved.")]
        public void Resolve_CircularDependency_ThrowsCircularDependencyException()
        {
            var container = new IoCContainer();
            container.Resolve<CircularDependency>();
        }

        [TestMethod]
        [ExpectedException(typeof(CircularDependencyException), "A class with a circular dependency was trying to be resolved.")]
        public void Resolve_CircularSetter_ThrowsCircularDependencyException()
        {
            var container = new IoCContainer();
            container.RegisterType<IMockService, MockService1>();
            container.Resolve<CircularSetter>();
        }

        [TestMethod]
        [ExpectedException(typeof(NotRegisteredException), "A class that wasn't registered was trying to be resolved.")]
        public void Resolve_Interface_NotRegistered_ThrowsNotRegisteredException()
        {
            var container = new IoCContainer();
            container.Resolve<InterfaceDependency>();
        }

        [TestMethod]
        [ExpectedException(typeof(NoAvailableConstructorException), "A class without constructors was trying to be resolved.")]
        public void Resolve_NoConstructor_ThrowsNoPublicConstructorException()
        {
            var container = new IoCContainer();
            container.Resolve<NoPublicConstructor>();
        }

    }
    #endregion Exceptions
}

File: Java/src/test/java/IoCContainerTest.java
import Exceptions.CircularDependencyException;
import Exceptions.NoAvailableConstructorException;
import Exceptions.NotRegisteredException;
import Mock.*;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.ExpectedException;

import java.util.concurrent.atomic.AtomicInteger;

import static org.junit.Assert.*;

/**
 * Created by user on 21/9/2015.
 */
public class IoCContainerTest {
    @Test
    public void resolve_NoDependencies_NotRegistered() {
        IoCContainer container = new IoCContainer();
        NoDependencies res = (NoDependencies) container.resolve(NoDependencies.class);

        assertNotNull(res);
    }

    @Test
    public void resolve_NoDependencies_Registered() {
        IoCContainer container = new IoCContainer();
        container.registerType(NoDependencies.class, NoDependencies.class);
        NoDependencies res = (NoDependencies) container.resolve(NoDependencies.class);

        assertNotNull(res);
    }

    @Test
    public void resolve_MultipleDependencies_NotRegistered() {
        IoCContainer container = new IoCContainer();
        MultipleDependencies res = (MultipleDependencies) container.resolve(MultipleDependencies.class);

        assertNotNull(res);
        assertNotNull(res.service1);
        assertNotNull(res.service2);
    }

    @Test
    public void resolve_Interface_Registered() {
        IoCContainer container = new IoCContainer();
        container.registerType(IMockService.class, MockService1.class);
        InterfaceDependency res = (InterfaceDependency) container.resolve(InterfaceDependency.class);

        assertNotNull(res.service);
        assertTrue(res.service instanceof MockService1);
    }

    @Test
    public void resolve_MultipleDependencies_AndInterface() {
        IoCContainer container = new IoCContainer();
        container.registerType(IMockService.class, MockService1.class);
        MultipleDependenciesAndInterface res = (MultipleDependenciesAndInterface) container.resolve(MultipleDependenciesAndInterface.class);

        assertNotNull(res.interfaceService);
        assertTrue(res.interfaceService instanceof MockService1);
        assertNotNull(res.service2);
        assertNotNull(res.service1);
    }

    @Test
    public void resolve_Instance() {
        IoCContainer container = new IoCContainer();
        container.registerInstance(container.resolve(NonStaticSingleton.class));

        container.resolve(NonStaticSingleton.class);
        assertEquals(1, NonStaticSingleton.nConstructed);
        container.resolve(NonStaticSingleton.class);
        assertEquals(1, NonStaticSingleton.nConstructed);
        container.resolve(NonStaticSingleton.class);
        assertEquals(1, NonStaticSingleton.nConstructed);
    }

    @Test
    public void resolve_InjectionAttributeConstructor() {
        IoCContainer container = new IoCContainer();
        InjectConstructorMock res = (InjectConstructorMock) container.resolve(InjectConstructorMock.class);

        assertNull(res.interfaceService);
        assertNotNull(res.service1);
        assertNotNull(res.service2);
    }


    @Test
    public void resolve_InjectionAttributeProperty() {
        IoCContainer container = new IoCContainer();
        InjectPropertyMock res = (InjectPropertyMock) container.resolve(InjectPropertyMock.class);

        assertNull(res.getInterfaceService());
        assertNotNull(res.getService1());
        assertNotNull(res.getService2());
    }

    @Test
    public void resolve_Func() {
        IoCContainer container = new IoCContainer();
        AtomicInteger num = new AtomicInteger (0);

        container.registerFunc(NoDependencies.class, () ->
        {
            num.incrementAndGet();
            return new NoDependencies();
        });

        NoDependencies res1 = (NoDependencies) container.resolve(NoDependencies.class);
        assertNotNull(res1);
        assertEquals(1, num.get());
        NoDependencies res2 = (NoDependencies) container.resolve(NoDependencies.class);
        assertNotNull(res2);
        assertEquals(2, num.get());
        NoDependencies res3 = (NoDependencies) container.resolve(NoDependencies.class);
        assertNotNull(res3);
        assertEquals(3, num.get());
    }

    @Rule
    public final ExpectedException circularDependencyException = ExpectedException.none();
    @Rule
    public final ExpectedException notRegisteredException = ExpectedException.none();
    @Rule
    public final ExpectedException illegalArgumentException = ExpectedException.none();
    @Rule
    public final ExpectedException noAvailableConstructorException = ExpectedException.none();

    @Test
    public void Resolve_NullFunc()
    {
        IoCContainer container = new IoCContainer();
        illegalArgumentException.expect(IllegalArgumentException.class);
        container.registerFunc(Object.class, null);
    }

    @Test
    public void resolve_NullInstance() {
        IoCContainer container = new IoCContainer();
        illegalArgumentException.expect(IllegalArgumentException.class);
        container.registerInstance(null);
    }

    @Test
    public void resolve_CircularDependency_ThrowsCircularDependencyException() {
        IoCContainer container = new IoCContainer();
        circularDependencyException.expect(CircularDependencyException.class);
        container.resolve(CircularDependency.class);
    }

    @Test
    public void Resolve_CircularDependency_ThrowsCircularDependencyException()
    {
        IoCContainer container = new IoCContainer();
        circularDependencyException.expect(CircularDependencyException.class);
        container.resolve(CircularDependency.class);
    }

    @Test
    public void Resolve_CircularSetter_ThrowsCircularDependencyException()
    {
        IoCContainer  container = new IoCContainer();
        container.registerType(IMockService.class, MockService1.class);
        circularDependencyException.expect(CircularDependencyException.class);
        container.resolve(CircularSetter.class);
    }


    @Test
    public void resolve_Interface_NotRegistered_ThrowsNotRegisteredException() {
        IoCContainer container = new IoCContainer();
        notRegisteredException.expect(NotRegisteredException.class);
        container.resolve(InterfaceDependency.class);
    }

    @Test
    public void resolve_NoConstructor_ThrowsNoPublicConstructorException() {
        IoCContainer container = new IoCContainer();
        noAvailableConstructorException.expect(NoAvailableConstructorException.class);
        container.resolve(NoPublicConstructor.class);
    }

}
